src/cpu/x64/utils/jit_io_helper.hpp

```diff
@@ -247,6 +247,27 @@ class jit_io_multi_dt_helper_t {
             storage_;
 };
 
+template <>
+void jit_io_helper_t<Xbyak::Zmm>::emu_gather(const Xbyak::Reg64 &src_reg,
+        const Xbyak::Zmm &indices_vmm, const Xbyak::Zmm &dst_vmm,
+        const bool tail);
+template <>
+void jit_io_helper_t<Xbyak::Ymm>::emu_gather(const Xbyak::Reg64 &src_reg,
+        const Xbyak::Ymm &indices_vmm, const Xbyak::Ymm &dst_vmm,
+        const bool tail);
+template <>
+void jit_io_helper_t<Xbyak::Xmm>::emu_gather(const Xbyak::Reg64 &src_reg,
+        const Xbyak::Xmm &indices_vmm, const Xbyak::Xmm &dst_vmm,
+        const bool tail);
+
+extern template class jit_io_helper_t<Xbyak::Zmm>;
+extern template class jit_io_helper_t<Xbyak::Ymm>;
+extern template class jit_io_helper_t<Xbyak::Xmm>;
+
+extern template class jit_io_multi_dt_helper_t<Xbyak::Zmm>;
+extern template class jit_io_multi_dt_helper_t<Xbyak::Ymm>;
+extern template class jit_io_multi_dt_helper_t<Xbyak::Xmm>;
```

Comment:
It's fine to have these `extern`s. Can't see why upper declarations needed... could you explain, please?

Reply:
The declarations of the explicit specializations are needed because it is ill-formed (no diagnostic required) for a user to see the primary and instantiate implicitly a specialization of the primary when a different part of the program defines an explicit specialisation: https://eel.is/c++draft/temp.spec#temp.expl.spec-7.sentence-1

Reply:
What I meant is `emu_gather` is a private method, and compiler should be able to resolve internals of same class correctly without additional declarations.

Following this logic, all other methods should be declared, but they are not. This seems to be inconsistent...

Reply:
The other member function templates are specialized from their primary, so there's no mismatch potential.

Ah, the explicit specialization declarations are now required because of the explicit instantiation declarations. Without the former, the latter cause an error:

```src/cpu/x64/utils/jit_io_helper.cpp:185:35: error: explicit specialization of 'emu_gather' after instantiation
void jit_io_helper_t<Xbyak::Zmm>::emu_gather(const Xbyak::Reg64 &src_reg,
                                  ^
src/cpu/x64/utils/jit_io_helper.hpp:242:23: note: explicit instantiation first required here
extern template class jit_io_helper_t<Xbyak::Zmm>;
                      ^
```

Once you _declare_ the explicit instantiation, that declares instantiations of all member templates, and once you've done that, you can't then go back and add an explicit specialization for one of them. That's why we _first_ have to declare the explicit specializations, and then the explicit instantiation takes care of everything else (but now excluding the already explicitly-specialized members).

Reply:
Alright, I was able to reproduce it. It was worth mentioning that `emu_gather` warning is a product of adding `extern template class jit_io_helper_t<Xbyak::Zmm>` in the first place.

I'm not a big fan of adding declarations for each specialization since it becomes clumsy after several of them. They may grow over time. In this particular case it feels like the code for `emu_gather` may be merged into a single instantiation without specializations, but it would be huge from me to ask you to do it.

I'm fine to leave declarations for now, but please leave a `TODO: remove specializations of `emu_gather` and these declaration by merging three specs into a single piece` comment above them.

I've also noticed some more errors with `store_byte_by_byte` and `load_byte_by_byte`.
Please add following patch to resolve those:
```
diff --git a/src/cpu/x64/utils/jit_io_helper.cpp b/src/cpu/x64/utils/jit_io_helper.cpp
index 70ba9d26d4..464646b2c5 100644
--- a/src/cpu/x64/utils/jit_io_helper.cpp
+++ b/src/cpu/x64/utils/jit_io_helper.cpp
@@ -553,16 +553,13 @@ void jit_io_helper_t<Vmm>::load(const Xbyak::Address &src_addr,
     }
 }

-template <>
-void jit_io_helper_t<Xbyak::Zmm>::load_byte_by_byte(
-        const Xbyak::Address &src_addr, const Xbyak::Zmm &dst_vmm,
-        const int load_size) {
-    assert("Load byte by byte is not supported for Zmms.");
-}
-
 template <typename Vmm>
 void jit_io_helper_t<Vmm>::load_byte_by_byte(const Xbyak::Address &src_addr,
         const Vmm &dst_vmm, const int load_size) {
+    static constexpr bool is_zmm = std::is_same<Vmm, Xbyak::Zmm>::value;
+    UNUSED(is_zmm);
+    assert(!is_zmm && "Load byte by byte is not supported for Zmms.");
+
     host_->uni_vxorps(dst_vmm, dst_vmm, dst_vmm);
     host_->load_data(data_type_, dst_vmm, src_addr, load_size);

@@ -706,15 +703,13 @@ void jit_io_helper_t<Vmm>::saturate(const Vmm &vmm) {
     host_->uni_vcvtps2dq(vmm, vmm);
 }

-template <>
-void jit_io_helper_t<Xbyak::Zmm>::store_byte_by_byte(const Xbyak::Zmm &src_zmm,
-        const Xbyak::Address &dst_addr, const int store_size) {
-    assert("Store byte by byte is not supported for Zmms.");
-}
-
 template <typename Vmm>
 void jit_io_helper_t<Vmm>::store_byte_by_byte(const Vmm &src_vmm,
         const Xbyak::Address &dst_addr, const int store_size) {
+    static constexpr bool is_zmm = std::is_same<Vmm, Xbyak::Zmm>::value;
+    UNUSED(is_zmm);
+    assert(!is_zmm && "Store byte by byte is not supported for Zmms.");
+
     const bool is_i8 = utils::one_of(data_type_, data_type::s8, data_type::u8);
     const bool is_xf16
             = utils::one_of(data_type_, data_type::bf16, data_type::f16);
```

Thanks.

Reply:
Sure!

Could we make those `assert`s into `static_assert`ions, though?

Reply:
```
../src/cpu/x64/utils/jit_io_helper.cpp:560:5: error: static_assert failed due to requirement '!std::is_same<Xbyak::Zmm, Xbyak::Zmm>::value' "Load byte by byte is not supported for Zmms."
    static_assert(!std::is_same<Vmm, Xbyak::Zmm>::value,
    ^             ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
../src/cpu/x64/utils/jit_io_helper.cpp:964:16: note: in instantiation of member function 'dnnl::impl::cpu::x64::io::jit_io_helper_t<Xbyak::Zmm>::load_byte_by_byte' requested here
template class jit_io_helper_t<Xbyak::Zmm>;
               ^
```
It seems unlikely to me...

Reply:
OK, shame! Using your original patch now.